File: spring_java/src/test/java/dev/pollito/spring_java/config/security/userdetails/SakilaUserDetailsServiceTest.java
package dev.pollito.spring_java.config.security.userdetails;

import static java.util.Optional.empty;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.when;

import dev.pollito.spring_java.sakila.staff.domain.model.Staff;
import dev.pollito.spring_java.sakila.staff.domain.port.out.StaffRepository;
import java.util.Optional;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.userdetails.UsernameNotFoundException;

@ExtendWith(MockitoExtension.class)
class SakilaUserDetailsServiceTest {

  @Mock private StaffRepository staffRepository;

  @InjectMocks private SakilaUserDetailsService sakilaUserDetailsService;

  @Test
  void loadUserByUsername() {
    Staff staff =
        Staff.builder()
            .id(1)
            .username("Mike")
            .password("encoded")
            .firstName("Mike")
            .lastName("Hillyer")
            .active(true)
            .build();
    when(staffRepository.findByUsername("Mike")).thenReturn(Optional.of(staff));

    SakilaUserDetails userDetails =
        (SakilaUserDetails) sakilaUserDetailsService.loadUserByUsername("Mike");

    assertEquals("Mike", userDetails.getUsername());
    assertEquals("encoded", userDetails.getPassword());
    assertTrue(userDetails.isAccountNonLocked());
    assertEquals(staff, userDetails.staff());
  }

  @Test
  void loadUserByUsernameNotFound() {
    when(staffRepository.findByUsername("Unknown")).thenReturn(empty());
    assertThrows(
        UsernameNotFoundException.class,
        () -> sakilaUserDetailsService.loadUserByUsername("Unknown"));
  }
}

File: spring_kotlin/src/test/kotlin/dev/pollito/spring_kotlin/config/security/userdetails/SakilaUserDetailsServiceTest.kt
package dev.pollito.spring_kotlin.config.security.userdetails

import dev.pollito.spring_kotlin.sakila.staff.domain.model.Staff
import dev.pollito.spring_kotlin.sakila.staff.domain.port.out.StaffRepository
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import java.util.Optional
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.security.core.userdetails.UsernameNotFoundException

@ExtendWith(MockKExtension::class)
class SakilaUserDetailsServiceTest {

  @MockK private lateinit var staffRepository: StaffRepository

  @InjectMockKs private lateinit var sakilaUserDetailsService: SakilaUserDetailsService

  companion object {
    private val STAFF =
        Staff(
            id = 1,
            firstName = "Mike",
            lastName = "Hillyer",
            username = "Mike",
            password = "password",
            email = "Mike.Hillyer@sakilastaff.com",
            active = true,
        )
  }

  @Test
  fun `loadUserByUsername returns user details`() {
    every { staffRepository.findByUsername("Mike") } returns Optional.of(STAFF)

    val result = sakilaUserDetailsService.loadUserByUsername("Mike")

    assertEquals("Mike", result.username)
    assertEquals("password", result.password)
    assertTrue(result.isAccountNonLocked)
    assertEquals(STAFF, (result as SakilaUserDetails).staff)
    verify { staffRepository.findByUsername("Mike") }
  }

  @Test
  fun `loadUserByUsername throws when staff not found`() {
    every { staffRepository.findByUsername("Unknown") } returns Optional.empty()

    assertFailsWith<UsernameNotFoundException> {
      sakilaUserDetailsService.loadUserByUsername("Unknown")
    }
  }
}
